java/component/keyple-plugin/remote-se/core/src/test/java/org/eclipse/keyple/plugin/remotese/core/impl/KeypleClientAsyncNodeImplTest.java

```diff
@@ -0,0 +1,651 @@
+        }
+        if (whenSendMessage) {
+            doAnswer(new Answer() {
+                @Override
+                public Object answer(InvocationOnMock invocation) throws Throwable {
+                    node.onError(sessionId, error);
+                    return null;
+                }
+            }).when(endpoint).sendMessage(msg);
+        }
+        if (whenCloseSession) {
+            doAnswer(new Answer() {
+                @Override
+                public Object answer(InvocationOnMock invocation) throws Throwable {
+                    node.onError(sessionId, error);
+                    return null;
+                }
+            }).when(endpoint).closeSession(sessionId);
+        }
+    }
+
+    void setEndpointError(boolean whenOpenSession, boolean whenSendMessage,
+            boolean whenCloseSession) {
+        if (whenOpenSession) {
+            doThrow(new RuntimeException()).when(endpoint).openSession(sessionId);
+        }
+        if (whenSendMessage) {
+            doThrow(new RuntimeException()).when(endpoint).sendMessage(msg);
+        }
+        if (whenCloseSession) {
+            doThrow(new RuntimeException()).when(endpoint).closeSession(sessionId);
+        }
+    }
+
+    @Before
+    public void setUp() {
+        super.setUp();
+        endpoint = mock(KeypleClientAsync.class);
+        node = new KeypleClientAsyncNodeImpl(handler, endpoint, 1);
+    }
+
+    @Test
+    public void openSession_whenOk_shouldCallEndpointAndReturn() {
+        setEndpointAnswer(true, false, false);
+        node.openSession(sessionId);
+        verify(endpoint).openSession(sessionId);
+        verifyNoMoreInteractions(endpoint);
+        verifyZeroInteractions(handler);
+    }
+
+    @Test(expected = KeypleTimeoutException.class)
+    public void openSession_whenTimeout_shouldThrowKeypleTimeoutException() {
+        setEndpointAnswer(false, false, true);
+        node.openSession(sessionId);
+    }
+
+    @Test(expected = RuntimeException.class)
+    public void openSession_whenEndpointError_shouldThrowEndpointError() {
+        setEndpointAnswer(false, false, true);
+        setEndpointError(true, false, false);
+        node.openSession(sessionId);
+    }
+
+    @Test(expected = IllegalArgumentException.class)
+    public void onOpen_whenSessionIdIsNull_shouldThrowIAE() {
+        node.onOpen(null);
+    }
+
+    @Test(expected = IllegalArgumentException.class)
+    public void onOpen_whenSessionIdIsEmpty_shouldThrowIAE() {
+        node.onOpen("");
+    }
+
+    @Test
+    public void onOpen_whenSessionIdIsUnknown_shouldDoNothing() {
+        node.onOpen(sessionIdUnknown);
+    }
+
+    @Test(expected = IllegalStateException.class)
+    public void onOpen_whenBadUse_shouldThrowISE() {
+        setEndpointAnswer(true, false, false);
+        openSessionSuccessfully(sessionId);
+        node.onOpen(sessionId);
+    }
+
+    @Test
+    public void onOpen_whenOkInThread1_shouldEndOpenSession() {
```

Comment:
duplicate of openSession_whenOk_shouldCallEndpointAndReturn

Reply:
The code is effectively duplicate but the behaviour tested is different.
The first test the openSession method behaviour, and the second one test the onOpen method behaviour on a specific context.